src/NuGet.Clients/PackageManagement.VisualStudio/Migrate/ProjectJsonToPackageRefMigrator.cs

```diff
@@ -11,35 +11,50 @@
 using System.Linq;
 using System.Threading.Tasks;
 using Task = System.Threading.Tasks.Task;
+using Microsoft.VisualStudio.Threading;
 using NuGet.PackageManagement.UI;
+using Microsoft.VisualStudio.Shell;
 
 namespace NuGet.PackageManagement.VisualStudio
 {
     public static class ProjectJsonToPackageRefMigrator
     {
         public static async Task MigrateAsync(LegacyCSProjPackageReferenceProject project, string dteProjectFullName)
         {
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
@jainaashish  [Threading rule no. 1](https://github.com/Microsoft/vs-threading/blob/master/doc/threading_rules.md#1-if-a-method-has-certain-thread-apartment-requirements-sta-or-mta-it-must-either) requires that async methods get themselves onto the thread they require. Only synchronous methods should throw if not called on the expected thread.

+1 to @rohit21agrawal: switching to the main thread is a very fast no-op if you're already there. 

Also, I don't see a JTF.RunAsync in this method. But I see several places in this method where the main thread is required. The safest (and usually best) approach is to stay on the UI thread unless you're absolutely sure the work you are doing doesn't require the UI thread, and then you only switch off for that small part. @jainaashish your default seems to be "stay off the UI thread except when you need it" but my experience from many partner teams is that that's a dangerous default that leads to hangs and crashes. You can still achieve great responsiveness, and higher reliability, by defaulting the opposite way.